app_dart/lib/src/request_handlers/refresh_cirrus_status.dart

```diff
@@ -37,60 +42,91 @@ class RefreshCirrusStatus extends ApiRequestHandler<Body> {
+    await for (Branch branch in branches) {
+      if (!regExps
+          .any((String regExp) => RegExp(regExp).hasMatch(branch.name))) {
+        continue;
       }
+      await for (FullTask task in datastore.queryRecentTasks(
+          taskName: 'cirrus', commitLimit: 15, branch: branch.name)) {
+        final String sha = task.commit.sha;
+        final String existingTaskStatus = task.task.status;
 
-      String newTaskStatus;
-      if (statuses.isEmpty) {
-        newTaskStatus = Task.statusNew;
-      } else if (statuses.any(kCirrusFailedStates.contains)) {
-        newTaskStatus = Task.statusFailed;
-        task.task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
-      } else if (statuses.any(kCirrusInProgressStates.contains)) {
-        newTaskStatus = Task.statusInProgress;
-      } else {
-        newTaskStatus = Task.statusSucceeded;
-        task.task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
-      }
+        log.debug(
+            'Found Cirrus task for branch ${branch.name}, commit $sha, with existing status $existingTaskStatus');
+
+        final List<String> statuses = <String>[];
+        const String name = 'flutter';
+        final List<CirrusResult> cirrusResults =
+            await queryCirrusGraphQL(sha, client, log, name);
+
+        /// Multiple branches may exist for same commit.
+        /// Update only when branches match
+        if (!cirrusResults.any((CirrusResult cirrusResult) =>
+            cirrusResult.branch == task.commit.branch)) {
+          continue;
+        }
+        for (Map<String, dynamic> runStatus in cirrusResults
+            .singleWhere((CirrusResult cirrusResult) =>
+                cirrusResult.branch == task.commit.branch)
+            .tasks) {
+          final String status = runStatus['status'] as String;
+          final String taskName = runStatus['name'] as String;
+          log.debug('Found Cirrus build status for $sha: $taskName ($status)');
+          statuses.add(status);
+        }
+
+        String newTaskStatus;
+        if (statuses.isEmpty) {
+          newTaskStatus = Task.statusNew;
+        } else if (statuses.any(kCirrusFailedStates.contains)) {
+          newTaskStatus = Task.statusFailed;
+          task.task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
+        } else if (statuses.any(kCirrusInProgressStates.contains)) {
+          newTaskStatus = Task.statusInProgress;
+        } else {
+          newTaskStatus = Task.statusSucceeded;
+          task.task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
+        }
 
-      if (newTaskStatus != existingTaskStatus) {
-        task.task.status = newTaskStatus;
-        await config.db.withTransaction<void>((Transaction transaction) async {
-          transaction.queueMutations(inserts: <Task>[task.task]);
-          await transaction.commit();
-        });
+        if (newTaskStatus != existingTaskStatus) {
+          task.task.status = newTaskStatus;
+          await config.db
```

Comment:
Maybe add the retries wrapper? everything related with datastore has a high failure rate.

Reply:
Added.